### PR TITLE
Python: further migrate packages from 2 to 3

### PR DESCRIPTION
**Pull request by FRidh**
###### Motivation for this change

Goal is to get rid of the `python` attribute.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 131719` at 238fa78d run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>14 packages marked as broken and skipped:</summary>
  <ul>
    <li>python38Packages.ignite</li>
    <li>python38Packages.pytorchWithCuda</li>
    <li>python38Packages.spacy</li>
    <li>python38Packages.spacy-transformers</li>
    <li>python38Packages.tensorflowWithCuda</li>
    <li>python38Packages.textacy</li>
    <li>python39Packages.ignite</li>
    <li>python39Packages.pytorchWithCuda</li>
    <li>python39Packages.spacy</li>
    <li>python39Packages.spacy-transformers</li>
    <li>python39Packages.tensorflowWithCuda</li>
    <li>python39Packages.textacy</li>
    <li>rocm-opencl-icd</li>
    <li>rocm-opencl-runtime</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/f214fc512019da9fe9863768a31f47d6">appleseed</a></li>
    <li><a href="https://gist.github.com/547adbc7f75f0cbbb2a9108beebd48f5">osl</a></li>
  </ul>
</details>
<details>
  <summary>532 packages skipped due to time constraints:</summary>
  <ul>
    <li>ansible (ansible_2_11 ,python39Packages.ansible-core)</li>
    <li>ansible-lint (python39Packages.ansible-lint)</li>
    <li>ansible_2_10 (python39Packages.ansible-base)</li>
    <li>apacheHttpdPackages.mod_tile (apacheHttpdPackages_2_4.mod_tile)</li>
    <li>appdaemon</li>
    <li>apprise (python39Packages.apprise)</li>
    <li>azure-cli</li>
    <li>beancount (python39Packages.beancount)</li>
    <li>beets</li>
    <li>beetsExternalPlugins.alternatives</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>30 packages built successfully:</summary>
  <ul>
    <li>coccinelle</li>
    <li>flowblade</li>
    <li>hammer</li>
    <li>libsigrok</li>
    <li>libsigrok_0_3</li>
    <li>lkl</li>
    <li>lyx</li>
    <li>matrix-appservice-discord</li>
    <li>maxima</li>
    <li>maxima-ecl</li>
    <li>mergerfs-tools</li>
    <li>mlt</li>
    <li>moosefs</li>
    <li>neard</li>
    <li>python38Packages.aprslib</li>
    <li>python38Packages.mlt</li>
    <li>python38Packages.mox3</li>
    <li>python38Packages.requests-mock</li>
    <li>python38Packages.subunit</li>
    <li>python38Packages.testrepository</li>
    <li>python39Packages.aprslib</li>
    <li>python39Packages.mlt</li>
    <li>python39Packages.mox3</li>
    <li>python39Packages.requests-mock</li>
    <li>python39Packages.sqlalchemy_migrate</li>
    <li>python39Packages.subunit</li>
    <li>python39Packages.testrepository</li>
    <li>sigrok-cli</li>
    <li>subunit</li>
    <li>yder</li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/name-and-version.md">name-and-version</a>

Did you mean to pass `pname` instead of `name` to `mkDerivation`?

Near pkgs/applications/science/math/maxima/default.nix:17:3:
```
   |
17 |   name = "${name}-${version}";
   |   ^
```

Near pkgs/applications/science/math/maxima/default.nix:16:10:
```
   |
16 |   inherit version;
   |          ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/development/python-modules/subunit/default.nix:17:20:
```
   |
17 |   inherit (subunit) name src meta;
   |                    ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl2` is a deprecated license, please check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/applications/science/math/maxima/default.nix:106:5:
```
    |
106 |     license = lib.licenses.gpl2;
    |     ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

File: pkgs/development/libraries/mapnik/default.nix [hunk 1]
```diff
@@ -1,6 +1,6 @@
 { lib, stdenv, fetchzip
 , boost, cairo, freetype, gdal, harfbuzz, icu, libjpeg, libpng, libtiff
-, libwebp, libxml2, proj, python, sqlite, zlib
+, libwebp, libxml2, proj, python3, python ? python3, sqlite, zlib
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/mapnik/default.nix:3]:
```suggestion
, libwebp, libxml2, proj, python3, sqlite, zlib
```

Comment (FRidh) [reply pkgs/development/libraries/mapnik/default.nix:3]: This is to separate the python used as build tool from that for which the binding is created.

Comment (SuperSandro2000) [reply pkgs/development/libraries/mapnik/default.nix:3]: oh, maybe we should add a comment for this. I am probably not the last one who stumbles upon this.

**Review by jonringer — APPROVED** [review]
LGTM

**Comment by collares** [comment]
In `pkgs/applications/science/math/sage/default.nix`, we do 

```
  maxima-ecl = pkgs.maxima-ecl.override {
    ecl = pkgs.ecl.override {
      useBoehmgc = true;
    };
  };
```

`ecl` adds `boehmgc` to its `propagatedBuildInputs` when `useBoehmgc == true`. Unfortunately, this overridden version of `maxima-ecl` does not work anymore. Vanilla `maxima-ecl`, as expected, uses `ecl` to compile Lisp files. When `ecl` was compiled with `useBoehmgc == true`, this is done by generating C files that use boehmgc's headers and compiling those files with `gcc`. Since this PR landed, this last step fails to find the right headers:

```
;;;
;;; Compiling /build/maxima-5.44.0/src/maxima-package.lisp.
;;; OPTIMIZE levels: Safety=2, Space=0, Speed=3, Debug=2
;;;
;;; End of Pass 1.
;;; Internal error:
;;;   ** Error code 1 when executing
;;; (EXT:RUN-PROGRAM "gcc" ("-I." "-I/nix/store/v8k8wdlgcx6dl4a5v1rplsy89dyhlndj-ecl-21.2.1/include/" "-I/nix/store/sipixdr4j94crffvl069mi94v7kyvqzv-gmp-6.2.1-dev/include" "-I/nix/store/w0v6cyzyhws18dahnv94bb0l0h6yf7zv-libffi-3.4.2-dev/include" "-D_GNU_SOURCE" "-D_FILE_OFFSET_BITS=64" "-g" "-O2" "-fPIC" "-Dlinux" "-O2" "-c" "binary-ecl/maxima-package.c" "-o" "binary-ecl/maxima-package.o")):
;;; In file included from /nix/store/v8k8wdlgcx6dl4a5v1rplsy89dyhlndj-ecl-21.2.1/include/ecl/ecl.h:37,
;;;                  from /nix/store/v8k8wdlgcx6dl4a5v1rplsy89dyhlndj-ecl-21.2.1/include/ecl/ecl-cmp.h:36,
;;;                  from binary-ecl/maxima-package.c:5:
;;; /nix/store/v8k8wdlgcx6dl4a5v1rplsy89dyhlndj-ecl-21.2.1/include/ecl/config.h:79:11: fatal error: gc/gc.h: No such file or directory
;;;    79 | # include "gc/gc.h"
;;;       |           ^~~~~~~~~
;;; compilation terminated.
```

Adding `lisp-compiler` to `maxima`'s `buildInputs` as well as `nativeBuildInputs` makes things succeed but does not feel like a correct solution. What is a good way to make `maxima-ecl` with `useBoehmgc == true` automatically propagate the `boehmgc` dependency to `ecl`'s users? If anyone wants to reproduce this problem, just try to build the `sage` derivation.

**Comment by collares** [comment]
Okay, I think #132269 fixes the above problem :)

**Comment by arkivm** [comment]
appleseed.studio still depends on [python2](https://github.com/appleseedhq/appleseed/blob/015adb503af58cb80103e0c3ddeefc20d99d204f/src/appleseed.studio/CMakeLists.txt#L37) and there is an open issue for [python3](https://github.com/appleseedhq/appleseed/issues/2817) support.
However, freebsd ports successfully migrated to python3 with these [set of patches](https://svnweb.freebsd.org/ports/head/graphics/appleseed/files/). It's a bit non-trivial though. Tagging appleseed maintainer @hodapp

**Pull request merged**